File: python/datashaper/datashaper/engine/verbs/unroll.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#
"""Unroll verb implementation."""

from typing import Any, cast

import pandas as pd

from datashaper.engine.create_verb_result import create_verb_result
from datashaper.engine.types import VerbResult
from datashaper.engine.verbs.verb_input import VerbInput
from datashaper.engine.verbs.verbs_mapping import verb
from datashaper.verbs import unroll


@verb(name="unroll", treats_input_tables_as_immutable=True)
def unroll_verb(input: VerbInput, **kwargs: Any) -> VerbResult:
    """Unroll a column."""
    result = unroll(cast(pd.DataFrame, input.get_input()), **kwargs)
    return create_verb_result(result)

File: javascript/workflow/src/verbs/unroll.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
import type { UnrollArgs } from '@datashaper/schema'

import type { ColumnTableStep } from './util/factories.js'
import { stepVerbFactory } from './util/factories.js'

export const unrollStep: ColumnTableStep<UnrollArgs> = (input, { columns }) =>
	input.unroll(columns)

export const unroll = stepVerbFactory(unrollStep)
